volatility3/framework/plugins/linux/lsblk.py

```diff
@@ -0,0 +1,388 @@
+        val = self._major(devt) + self._minor(devt)
+
+        # Need these next 4 values for this hash function: https://elixir.bootlin.com/linux/v5.4/source/fs/inode.c#L474
+        i_hash_mask = vmlinux.object_from_symbol("i_hash_mask")
+        i_hash_shift = vmlinux.object_from_symbol("i_hash_shift")
+
+        # in linux 4.7, the GOLDEN_RATIO_PRIME constant changes. This struct gains a member from 4.6 to 4.7
+        if vmlinux.get_type("file_operations").has_member("iterate_shared"):
+            GOLDEN_RATIO_PRIME = GOLDEN_RATIO_PRIME_AFTER_4_7
+        else:
+            GOLDEN_RATIO_PRIME = GOLDEN_RATIO_PRIME_BEFORE_4_7
+
+        L1_CACHE_BYTES = 1 << 6
+
+        # Previously mentioned hash function
+        val = (val * blockdev_superblock) ^ (GOLDEN_RATIO_PRIME + val) // L1_CACHE_BYTES
+        val = val ^ ((val ^ GOLDEN_RATIO_PRIME) >> i_hash_shift)
+        val = val & i_hash_mask
+
+        bucket_size = vmlinux.get_type("hlist_head").size
+        hlist_head_offset = inode_hashtable + (val * bucket_size)
+
+        hlist_head = vmlinux.object(
+            object_type="hlist_head",
+            offset=hlist_head_offset,
+            absolute=True,
+        )
+        hlist_node = hlist_head.first
+        if not hlist_node:
+            return 0
+        while hlist_node:
+            inode = linux.LinuxUtilities.container_of(hlist_node, "inode", "i_hash", vmlinux)
+            # These checks are from the find_inode function https://elixir.bootlin.com/linux/v5.4/source/fs/inode.c#L805
+            # Trying to find the inode with the correct device number and calculated inode number from the inode cache
+            if inode.i_sb == blockdev_superblock and inode.i_rdev == devt:
+                break
+            hlist_node = hlist_node.next
+
+        return inode.i_bdev
+
+    def _device_iterator(self, context, klist_devices):
+        vmlinux = context.modules[self.config["kernel"]]
+        # Linux v5.1 moves device->knode_class to device_private->knode_class
+        knode_class_in_private_device = vmlinux.get_type("device_private").has_member(
+            "knode_class"
+        )
+
+        if knode_class_in_private_device:
+            for klist_node in klist_devices.k_list.to_list(
+                vmlinux.symbol_table_name + constants.BANG + "klist_node", "n_node"
+            ):
+                device_private = linux.LinuxUtilities.container_of(klist_node.vol.offset, "device_private", "knode_class", vmlinux)
+                yield device_private.device.dereference()
+        else:
+            for klist_node in klist_devices.k_list.to_list(
+                vmlinux.symbol_table_name + constants.BANG + "klist_node", "n_node"
+            ):
+                device = linux.LinuxUtilities.container_of(klist_node.vol.offset, "device", "knode_class", vmlinux)
+                yield device
+
+    def _get_name(self, vmlinux, block_device, gendisk):
```

Comment:
Also, rather than passing in vmlinux, the convention for volatility plugins is to pass in the context as the first parameter and then the module name as the second, and then lookup the module in `context.modules`.  Sticking to the convention should make it easier if/when we come to parallelize things rather than passing around lots of objects, it's just the context and then a reference for something inside the context.